runtime/compiler/env/VMJ9.cpp

```diff
@@ -1294,6 +1294,69 @@ TR_J9VMBase::getObjectClassInfoFromObjectReferenceLocation(TR::Compilation *comp
    return ci;
    }
 
+TR::KnownObjectTable::ObjectInfo
+TR_J9VMBase::getObjClassInfoFromKnotIndex(TR::Compilation *comp, TR::KnownObjectTable::Index knotIndex)
+   {
+   TR::KnownObjectTable *knot = comp->getKnownObjectTable();
+   // Check to see whether the known object entry already holds the desired info.
+   // If so, return the cached version. Otherwise, fetch the info and cache it.
+   TR::KnownObjectTable::ObjectInfo existingObjInfo = knot->getObjectInfo(knotIndex);
+   TR::KnownObjectTable::ObjectInfo retrievedObjInfo;
+   TR::KnownObjectTable::ObjectInfo answerObjInfo;
+   TR_ASSERT_FATAL(existingObjInfo._jniReference, "The jniReference is missing for knowIndex %d", knotIndex);
+   bool infoIsCached = existingObjInfo._clazz != NULL;
+   if (comp->getOption(TR_DisableKnownObjectTableCaching) ||
+       !infoIsCached ||
+       comp->getOption(TR_EnableKnownObjectTableCachingVerification))
+      {
+      // Retrieve the data from the VM
+      TR::VMAccessCriticalSection vmAccess(comp);
+
+      uintptr_t objectReference = knot->getPointer(knotIndex);
+      retrievedObjInfo._jniReference = existingObjInfo._jniReference;
+      retrievedObjInfo._clazz = getObjectClass(objectReference);
+      retrievedObjInfo._isString = isString(retrievedObjInfo._clazz);
+      retrievedObjInfo._jlClass = getClassClassPointer(retrievedObjInfo._clazz);
```

Comment:
I don't think we should really need to store `_jlClass` for every known object. It doesn't really meaningfully depend on the `_clazz` here. `getClassClassPointer()` just needs any old class so that it can find an instance of `java/lang/Class`. I understand though that VP needs to know whether we had it at this point. (It's pretty ridiculous that we can't just always guarantee that the JIT knows what it is...) So maybe we should keep it there for now and leave sorting that out as future work. Maybe worth a TODO?

Reply:
It makes sense to reduce the number of fields that are cached as much as possible. Since this will require more testing my plan is to merge this PR ASAP, backport to upcoming release v0.57.0 (deadline on 2025/12/09), and immediately open another PR with proposed solution, but deliver just to master (dev) branch.